### ISSUE TITLE
Recombinations in the DTWF model

### ISSUE THREAD

**Issue by TPPSellinger** [tskit-dev/msprime#936]
Hi,
I seem to never have breakpoints in the "dtwf" model. Here is what I do (python3) :

import msprime
ts = msprime.simulator_factory(
                    Ne=10**4, recombination_rate=1e-7,length=10**7,
                    population_configurations=[
                    msprime.PopulationConfiguration(
                    sample_size=10, initial_size=10**4,
                    growth_rate=0)],
                    model="dtwf")
ts.run()
ts.num_breakpoints 

This value is always equal to 0. However, if I put mutations (mutation rate equal to recombination rate), there are segregating sites. Hence I woul expect to have some breakpoints. I am not familiar with this model, are those results normal or is there an error somewhere?

**Comment by petrelharp** [comment]
Well, this is odd - tree sequences don't have a `num_breakpoints`
attribute. However, doing `len(list(ts.breakpoints()))` gives me a nonzero
number when I do this:
 dts = msprime.simulate(10, recombination_rate=1e-7, length=1e7,
model='dtwf`)

... what did you actually do?

**Comment by jeromekelleher** [comment]
Yes, can you paste in the actual code please @TPPSellinger? You can use [code blocks](https://wordpress.com/support/markdown-quick-reference/) to format it it properly.

**Comment by TPPSellinger** [comment]
Sorry a typos got in there. Here is the actual code (python3) : 

import msprime
ts = msprime.simulator_factory(
Ne=10000, recombination_rate=1e-8,length=10000000,
population_configurations=[
msprime.PopulationConfiguration(
sample_size=10, initial_size=10000,
growth_rate=0)],
model="dtwf")

ts.run()
ts.num_breakpoints

it outputs 0. But you are right @petrelharp, when I do  :  

import msprime
ts = msprime.simulate(Ne=10000, recombination_rate=1e-8,length=10000000,population_configurations=[msprime.PopulationConfiguration(sample_size=10, initial_size=10000,growth_rate=0)],model="dtwf")
ts.get_num_trees()

I have something around 9000.  However, I am interested in the number of recombination event and not number of trees (which should be higher). I thought I could get the number of recombination event using .simulator_factory. I hope this can help.

**Comment by jeromekelleher** [comment]
Ah, I see the issue now. The DTWF simulator isn't recording the breakpoints in the same way as the Hudson-based simulators. It would be a good bit of work to address this, so I think the we just don't worry about this check for DTWF.

However I am concered about the segfault you mentioned in your email @TPPSellinger  - can you give more details here please?

ps. I edited your post above to use code blocks (see the link in my last post for details)

**Comment by TPPSellinger** [comment]
Thanks for editing my post. I only had the core dump error once. I tried to relaunch it many time since, but never the error occured again. Here is what I tried : 
~~~~
import msprime
ts = msprime.simulator_factory(Ne=10000, recombination_rate=1e-7,length=10000000,population_configurations=[msprime.PopulationConfiguration(sample_size=10, initial_size=10000,growth_rate=0)],model="dtwf")
ts.run()
ts = msprime.simulate(Ne=10000, recombination_rate=1e-7,mutation_rate=1e-7,length=10000000,population_configurations=[msprime.PopulationConfiguration(sample_size=10, initial_size=10000,growth_rate=0)],model="dtwf")
~~~~
As I was running it through the verification.py (cf verify_recombination) , I don't know in which one the error occured. In the pull request I did yesterday, you can uncomment line 2957 of the verification.py (but the test is very slow). Core dump error occured when running :  
~~~~
verifier.add_dtwf_recombination()
~~~~

**Comment by petrelharp** [comment]
Oh, I see - additionally, `num_breakpoints` is an attribute of `Simulator`s, not `TreeSequence`s. Seems like an oversight?

**Comment by jeromekelleher** [comment]
> Oh, I see - additionally, num_breakpoints is an attribute of Simulators, not TreeSequences. Seems like an oversight?

Like we should have ``num_breakpoints`` on the ts as well @petrelharp? Originally I was trying to keep the distinction between recombination breakpoints that happen in the simulations (that don't necessary generate trees) and the observed distinct trees. (I've updated @TPPSellinger's code to say ``sim.num_breakpoints`` to make it a bit clearer what's happening.)

**Comment by jeromekelleher** [comment]
I would like to reproduce this segfault though. Would you mind seeing if you can make it happen again please @TPPSellinger?

**Comment by TPPSellinger** [comment]
Yes. I will try to make it happen again and let you know once I managed.

**Comment by TPPSellinger** [comment]
I finally managed to get the core dump error again. Here is the error :
 
python3: lib/msprime.c:1581: msp_dtwf_generate_breakpoint: Assertion `k > start' failed.
Aborted (core dumped)

Here is the set of parameters (to put in verification.py) : 
~~~~
    def run_DTWF_recombinations(self):
        basedir = "tmp__NOBACKUP__/DTWF_recombinations"
        self.verify_recombination(
            basedir, f"n=100_dtwf",
            sample_size=100, Ne=10**4, r=1e-4,
            L=10**8, model="dtwf")
~~~~
I don't always have the error. You might need to run it couple of times before having it. In addition, this might be of importance. As I was exploring the parameter space leading to the error, I was running two verification.py (where all test are commented except verifier.add_dtwf_recombination() ) in parallel on my machine (I have 16 Gb of memory and 8 cores). I don't know if the error now becomes normal. Should I try to get it while running only one verification.py on my machine ? 

PS: I don't think I was running two verification.py on parallel the first time I had the error (but it's not impossible).

**Comment by jeromekelleher** [comment]
Excellent, thanks @TPPSellinger.

**Comment by jeromekelleher** [comment]
Closed in favour of #938. Thanks for the bug report @TPPSellinger.

**Comment by petrelharp** [comment]
> Originally I was trying to keep the distinction between recombination breakpoints that happen in the simulations (that don't necessary generate trees) and the observed distinct trees.

Oh, I get the distinction! No, the current scheme makes sense.
